src/librbd/crypto/CryptoObjectDispatch.cc

```diff
@@ -21,71 +22,376 @@
 namespace librbd {
 namespace crypto {
 
+using librbd::util::create_context_callback;
 using librbd::util::data_object_name;
 
 template <typename I>
-struct C_EncryptedObjectReadRequest : public Context {
-
+struct C_AlignedObjectReadRequest : public Context {
     I* image_ctx;
-    CryptoInterface* crypto;
+    ceph::ref_t<CryptoInterface> crypto;
     uint64_t object_no;
     io::ReadExtents* extents;
-    Context* onfinish;
+    IOContext io_context;
+    const ZTracer::Trace parent_trace;
+    uint64_t* version;
+    Context* on_finish;
     io::ObjectDispatchSpec* req;
+    bool disable_read_from_parent;
 
-    C_EncryptedObjectReadRequest(
-            I* image_ctx, CryptoInterface* crypto, uint64_t object_no,
-            io::ReadExtents* extents, IOContext io_context,
+    C_AlignedObjectReadRequest(
+            I* image_ctx, ceph::ref_t<CryptoInterface> crypto,
+            uint64_t object_no, io::ReadExtents* extents, IOContext io_context,
             int op_flags, int read_flags, const ZTracer::Trace &parent_trace,
             uint64_t* version, int* object_dispatch_flags,
-            Context** on_finish,
-            Context* on_dispatched) : image_ctx(image_ctx),
-                                      crypto(crypto),
-                                      object_no(object_no),
-                                      extents(extents),
-                                      onfinish(on_dispatched) {
-      *on_finish = util::create_context_callback<
-              Context, &Context::complete>(*on_finish, crypto);
+            Context* on_dispatched
+            ) : image_ctx(image_ctx), crypto(crypto), object_no(object_no),
+                extents(extents), io_context(io_context),
+                parent_trace(parent_trace), version(version),
+                on_finish(on_dispatched) {
+      disable_read_from_parent =
+              read_flags & io::READ_FLAG_DISABLE_READ_FROM_PARENT;
```

Comment:
Nit: just to keep it as a true boolean:

```suggestion
              ((read_flags & io::READ_FLAG_DISABLE_READ_FROM_PARENT) != 0);
```